homeassistant/components/gree/__init__.py

```diff
@@ -21,52 +31,61 @@ async def async_setup(hass: HomeAssistant, config: dict):
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Set up Gree Climate from a config entry."""
-    devices = []
+    gree_discovery = DiscoveryService(hass)
+    hass.data[DATA_DISCOVERY_SERVICE] = gree_discovery
 
-    # First we'll grab as many devices as we can find on the network
-    # it's necessary to bind static devices anyway
-    _LOGGER.debug("Scanning network for Gree devices")
+    @callback
+    def shutdown_event(_: Event) -> None:
+        if hass.data[DOMAIN].get(DATA_DISCOVERY_INTERVAL) is not None:
+            hass.data[DOMAIN][DATA_DISCOVERY_INTERVAL]()
+            hass.data[DOMAIN].pop(DATA_DISCOVERY_INTERVAL)
+
+        if hass.data.get(DATA_DISCOVERY_SERVICE) is not None:
+            del hass.data[DATA_DISCOVERY_SERVICE]
```

Comment:
This code is duplicated in `async_unload_entry`, do the cleanup in a common helper.

Reply:
The idea is to make sure that the discovery process isn't left around if there is any case where shutdown can be called, and HomeAssistant keeps running. But I am unsure if this is possible, given than do you think it should still be removed?

Reply:
Neither, discovery does not require a thread, and the event is unsubscribes in the `async_unload_entry` which I think covers the needed cases. The only thing this cleans up is the following lines:

```python
gree_discovery = DiscoveryService(hass)
hass.data[DATA_DISCOVERY_SERVICE] = gree_discovery
```

There are no side effects of not cleaning it up, other than having the object persist in memory.

Reply:
Actually I don't think this should be here at all, cleanup is handled in `async_unload_entry` anyway

Reply:
`async_unload_entry` is not called at shutdown, but you don't really need to worry about cleaning up at shutdown unless it:
- Is somehow required by the library to do a clean exit
- Not cleaning up prevents Home Assistant from shutting down, for example running threads

Reply:
Ok, I removed the event, it is unnecessary/